### PR TITLE
Merge ModuleViewAndEditFix into Integration

### PR DESCRIPTION
**Pull request by xaterz**
What has been changed in this branch:
* Fixed Issue #18 by changing Edit Module pages to use GET method instead of POST
* Fixed Issue #20 by adding input validation for both View Module and Edit Module (can catch invalid modules and AY-Sems)
* Added cancel button in Edit module
* Added Back to Module Listing button in module overview
* Moved pre-req and preclusions from specific info to general info
* Added 'Edit Specific Module Info' button in module overview. On click, will scroll user to tentative info table

### CODE REVIEW

File: components/handlers/module_edit.py [hunk 1]
```diff
@@ -8,61 +8,114 @@
...
-                    int(module_mc) != int(old_module_mc)):
-                model.store_original_module_info(module_code, old_module_name,
-                                                 old_module_desc, old_module_mc)
-                outcome = model.update_module(module_code, module_name, module_desc, module_mc)
+        outcome = True
+        module_info = model.get_module(module_code)
+        if module_info is None:
+            outcome = False
+        elif (module_name != old_module_name or module_desc != old_module_desc or
+                int(module_mc) != int(old_module_mc)):
```
Comment (helloqx) [inline components/handlers/module_edit.py]: Remove 2 spaces

**Comment by helloqx** [comment]
Some comments:
* Added Back to Module Listing button in module overview
>> If access View Module from other areas such as Modified Modules, the button still says Back to Module List.
* Added 'Edit Specific Module Info' button in module overview. On click, will scroll user to tentative info table
>> General Info > Edit Specific > Cancel jumps user to Specific Overview of that Sem instead of back to General Info

**Comment by xaterz** [comment]
To solve this, will need to track the page that the user has visited previously. This is not very viable.

**Comment by tgqiang** [comment]
Revised test cases in <code>test_view_module.py</code> and <code>test_edit_module_fixture.py</code>.

Former test case file passes, latter test case file fails. Failure in latter test case file due to missing quota validation check in back-end if front-end was bypassed somehow, and was triggered by attempting to set module quota to **-1**.

**Comment by tgqiang** [comment]
> - Added Back to Module Listing button in module overview
> If access View Module from other areas such as Modified Modules, the button still says Back to Module List.
> - Added 'Edit Specific Module Info' button in module overview. On click, will scroll user to tentative info table
> General Info > Edit Specific > Cancel jumps user to Specific Overview of that Sem instead of back to General Info

> To solve this, will need to track the page that the user has visited previously. This is not very viable.

@helloqx @xaterz I think that using web.ctx would resolve this issue.

File: components/handlers/module_edit.py [hunk 2]
```diff
@@ -8,61 +8,114 @@
...
     def POST(self):
         '''
-            Handles the loading and submission of the 'edit mounting' form
+            Handles the submission of the 'Edit Specific Module Info' page
         '''
         data = web.input()
-        form_status = data.status
         module_code = data.code
         ay_sem = data.aySem
         try:
```
Comment (tgqiang) [inline components/handlers/module_edit.py]: Case of exceptional quota values falling outside of permissible range is not handled here.

**Comment by helloqx** [comment]
>> Added Back to Module Listing button in module overview
If access View Module from other areas such as Modified Modules, the button still says Back to Module List.
Added 'Edit Specific Module Info' button in module overview. On click, will scroll user to tentative info table
General Info > Edit Specific > Cancel jumps user to Specific Overview of that Sem instead of back to General Info

>To solve this, will need to track the page that the user has visited previously. This is not very viable.
@helloqx @xaterz I think that using web.ctx would resolve this issue.

web.ctx is global, cookies more possible.
Not high priority tho

**Comment by xaterz** [comment]
>>@helloqx @xaterz I think that using web.ctx would resolve this issue.

> web.ctx is global, cookies more possible.
Not high priority tho

Ya I think it's not very necessary to specially customise the button for the user. The button is just there for the user's convenience, that's all.

**Comment by xaterz** [comment]
Added backend quota validation. All tests passed. @tgqiang @helloqx please review again and approve.

File: test/test_edit_module_fixture.py [hunk 3]
```diff
@@ -79,26 +134,29 @@ def test_module_edit_submit_response(self):
...
-        root = self.test_app.get(self.URL_INDIVIDUAL_MODULE_VIEW)
-        assert_equal(root.status, 200)
-        submit_button = root.forms__get()[self.EDIT_MODULE_BUTTON_FORM_NAME]
-        response = submit_button.submit()
+        root = self.test_app.get(self.URL_EDIT_MODULE_SPECIFIC_VALID)
+        edit_module_mounting_form = root.forms__get()["edit-mounting-form"]
+        edit_module_mounting_form.__setitem__("quota", -1)
+        response = edit_module_mounting_form.submit()
 
+        print(response)
```
Comment (helloqx) [inline test/test_edit_module_fixture.py]: is this a debug statement?

File: components/handlers/module_edit.py [hunk 4]
```diff
@@ -72,15 +125,18 @@ def POST(self):
...
-        elif form_status == "submit":
-            old_mounting_value = data.oldMountingValue
-            mounting_status = data.mountingStatus
 
+        module_info = model.get_module(module_code)
+        if module_info is None:
+            outcome = False
+        elif ay_sem not in self.list_of_future_ay_sems:
+            outcome = False
+        elif quota < 0 or (quota is not None and type(quota) != int):
```
Comment (helloqx) [inline components/handlers/module_edit.py]: pylint: Consider using typecheck

File: test/test_view_individual_module.py [hunk 5]
```diff
@@ -128,68 +127,18 @@ def test_view_individual_module_invalid_code_response(self):
...
-    #     NOTE: Checking for invalid module quota is not implemented
-    # '''
-    # '''
-    # @raises(Exception)
-    # def test_view_invalid_module_overview_invalid_quota_response(self):
-    #     # Quota used here is '70' (actual is '60')
-    #     # an exception WILL be encountered here
-    #     root = self.testApp.get(self.URL_CONTAIN_CODE_AY_INVALID_QUOTA)
-    # '''
+    '''
```
Comment (helloqx) [inline test/test_view_individual_module.py]: Shift docstring below def line for consistency

File: components/handlers/module_edit.py (hunk 4 above)
Comment (helloqx) [inline components/handlers/module_edit.py]: This line is causing all edits to fail!

**Pull request merged**